File: android/webview/src/main/kotlin/io/rynbridge/webview/WebViewModule.kt
package io.rynbridge.webview

import io.rynbridge.core.*

class WebViewModule(provider: WebViewProvider) : BridgeModule {

    override val name = "webview"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "open" to { payload ->
            val url = payload["url"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: url")
            val title = payload["title"]?.stringValue
            val style = payload["style"]?.stringValue ?: "modal"
            val allowedOrigins = payload["allowedOrigins"]?.arrayValue
                ?.mapNotNull { it.stringValue }
                ?: emptyList()
            val options = OpenOptions(
                url = url,
                title = title,
                style = style,
                allowedOrigins = allowedOrigins
            )
            val result = provider.open(options)
            result.toPayload()
        },
        "close" to { payload ->
            val webviewId = payload["webviewId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: webviewId")
            provider.close(webviewId)
            emptyMap()
        },
        "sendMessage" to { payload ->
            val targetId = payload["targetId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: targetId")
            val data = payload["data"]?.dictionaryValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: data")
            provider.sendMessage(targetId, data)
            emptyMap()
        },
        "postEvent" to { payload ->
            val targetId = payload["targetId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: targetId")
            val event = payload["event"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: event")
            val data = payload["data"]?.dictionaryValue
            provider.postEvent(targetId, event, data)
            emptyMap()
        },
        "getWebViews" to { _ ->
            val result = provider.getWebViews()
            result.toPayload()
        },
        "setResult" to { payload ->
            val data = payload["data"]?.dictionaryValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: data")
            provider.setResult(data)
            emptyMap()
        }
    )
}

File: ios/Sources/RynBridgeWebView/WebViewModule.swift
import Foundation
import RynBridge

public struct WebViewModule: BridgeModule, Sendable {
    public let name = "webview"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: WebViewProvider) {
        actions = [
            "open": { payload in
                guard let url = payload["url"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: url")
                }
                let title = payload["title"]?.stringValue
                let style = payload["style"]?.stringValue ?? "modal"
                let allowedOrigins: [String]
                if let arr = payload["allowedOrigins"]?.arrayValue {
                    allowedOrigins = arr.compactMap { $0.stringValue }
                } else {
                    allowedOrigins = []
                }
                let options = OpenOptions(
                    url: url,
                    title: title,
                    style: style,
                    allowedOrigins: allowedOrigins
                )
                let result = try await provider.open(options: options)
                return result.toPayload()
            },
            "close": { payload in
                guard let webviewId = payload["webviewId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: webviewId")
                }
                try await provider.close(webviewId: webviewId)
                return [:]
            },
            "sendMessage": { payload in
                guard let targetId = payload["targetId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: targetId")
                }
                guard let data = payload["data"]?.dictionaryValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: data")
                }
                try await provider.sendMessage(targetId: targetId, data: data)
                return [:]
            },
            "postEvent": { payload in
                guard let targetId = payload["targetId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: targetId")
                }
                guard let event = payload["event"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: event")
                }
                let data = payload["data"]?.dictionaryValue
                provider.postEvent(targetId: targetId, event: event, data: data)
                return [:]
            },
            "getWebViews": { _ in
                let result = try await provider.getWebViews()
                return result.toPayload()
            },
            "setResult": { payload in
                guard let data = payload["data"]?.dictionaryValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: data")
                }
                provider.setResult(data: data)
                return [:]
            },
        ]
    }
}

#if canImport(UIKit)
extension WebViewModule {
    public init() {
        self.init(provider: DefaultWebViewProvider())
    }
}
#endif
